plutarch-test/conditional/Plutarch/TryFromSpec.hs

```diff
@@ -0,0 +1,506 @@
+      "concatenate two lists, illegal (wrong elements in list)"
+        @| validator # pforgetData l1 # pforgetData l4 # validContext0 @-> pfails
+      "concatenate two lists, illegal (more than one output)"
+        @| validator # pforgetData l1 # pforgetData l2 # invalidContext1 @-> pfails
+    "example2" @\ do
+      "recovering a record succeeds"
+        @| recoverAB @-> psucceeds
+
+------------------- Checking deeply, shallowly and unwrapping ----------------------
+
+checkDeep ::
+  forall (target :: PType) (actual :: PType).
+  ( PTryFrom PData (PAsData target)
+  , PIsData actual
+  , PIsData target
+  ) =>
+  ClosedTerm (PAsData actual) ->
+  ClosedTerm (PAsData target)
+checkDeep t = unTermCont $ fst <$> TermCont (ptryFrom @(PAsData target) $ pforgetData t)
+
+checkDeepUnwrap ::
+  forall (target :: PType) (actual :: PType) (s :: S).
+  ( PTryFrom PData (PAsData target)
+  , PIsData actual
+  , PIsData target
+  ) =>
+  Term s (PAsData actual) ->
+  Term s (PAsData target)
+checkDeepUnwrap t = unTermCont $ fst <$> TermCont (ptryFrom @(PAsData target) $ pforgetData t)
+
+sampleStructure :: Term _ (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData PInteger)))))))
+sampleStructure = pdata $ psingleton #$ pdata $ psingleton #$ toDatadList [1 .. 100]
+
+-- | PData serves as the base case for recursing into the structure
+partialCheck :: Term _ (PAsData (PBuiltinList (PAsData (PBuiltinList PData))))
+partialCheck =
+  let dat :: Term _ PData
+      dat = pforgetData sampleStructure
+   in unTermCont $ fst <$> TermCont (ptryFrom dat)
+
+fullCheck :: Term _ (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData (PBuiltinList (PAsData PInteger)))))))
+fullCheck = unTermCont $ fst <$> TermCont (ptryFrom $ pforgetData sampleStructure)
+
+------------------- Example: untrusted Redeemer ------------------------------------
+
+newtype PNatural (s :: S) = PMkNatural (Term s PInteger)
+  deriving (PlutusType, PIsData, PEq, POrd) via (DerivePNewtype PNatural PInteger)
+
+-- | partial
+pmkNatural :: Term s (PInteger :--> PNatural)
+pmkNatural = plam $ \i -> pif (i #< 0) (ptraceError "could not make natural") (pcon $ PMkNatural i)
+
+newtype Flip f b a = Flip (f a b)
+
+instance Reducible (f a b) => Reducible (Flip f b a) where
+  type Reduce (Flip f b a) = Reduce (f a b)
+
+instance PTryFrom PData (PAsData PNatural) where
+  type PTryFromExcess PData (PAsData PNatural) = Flip Term PNatural
+  ptryFrom' opq = runTermCont $ do
+    (ter, exc) <- TermCont $ ptryFrom @(PAsData PInteger) opq
+    ver <- tcont $ plet $ pmkNatural #$ exc
+    pure $ (punsafeCoerce ter, ver)
+
+validator :: Term s PValidator
+validator = phoistAcyclic $
+  plam $ \dat red ctx -> unTermCont $ do
+    trustedRedeemer <- (\(snd -> red) -> red) <$> (TermCont $ ptryFrom @(PAsData (PBuiltinList (PAsData PNatural))) red)
+    let trustedDatum :: Term _ (PBuiltinList (PAsData PNatural))
+        trustedDatum = pfromData $ punsafeCoerce dat
```

Comment:
might be nice if we had a way to "verify" datums without `punsafeCoerce`. something that actually just does `punsafeCoerce` but is only restricted to datums. `PDatum` might actually come in handy for once.

Reply:
Not all datums are trusted.

Reply:
Some times you don't need to verify the redeemer either.